ffuf: add page (#6183)
tldr-pages/tldr: pages/linux/ffuf.md

# ffuf

> A fast web fuzzer written in Go.
> More information: <https://github.com/ffuf/ffuf>.

> - Directory discovery using a wordlist on a target website colorized and verbose:

Reviewer: What exactly does this search for? Also, example descriptions should always be in imperative form, e.g. "Find directories on a target website ...".

Author: addressed in 4fe26e86dfcc04de1115ef613c87693a5e0b7333

> - Directory discovery using a wordlist on a target website colorized and verbose:

Reviewer: All descriptions should be written in [`imperative`](https://en.wikipedia.org/wiki/Imperative_mood#English). For example:
```suggestion
- Find subdomains of a target website using a wordlist and show colorized and verbose output:
```

Author: I understand what you are getting at.  
However, the line 6 & 8 description/command pair is a directory discovery execution and not a subdomain discovery execution.
I've added in an extra command for subdomain discovery, so the current program description is more accurate.  This will take the number of examples to 6 from the recommend 5.

If the "discovery" word choice is poor, I am open to suggestions.
My current rationale is:
"Find" is more akin to locating something that you know exists and do not know where its current whereabouts are.
Whereas "discovery" is identifying or interacting with something that there was no previous knowledge of.

In my experience, the tool/program is used as a reconnaissance tool, on unfamiliar systems, hence the "discovery" stance.

addressed in 60d5192b7124796d1ce615d85a0cb1de0d18135e

Reviewer: Ah, okay. But my point was that the sentences should be written like `Discover directories using [...]`, instead of `Directory discovery [...]`, since that's the syntax we use in all pages. I can't really do that accurately enough since I don't understand the command well enough.

Author: Gotcha.
Edited the command descriptions to be in line with your comments.

addressed in 7564c8f458e27155ea9e882cf70b400f89895338

> `ffuf -w {{/path/to/wordlist}} -u {{https://target/FUZZ}} -c -v`

Reviewer: If the path to the wordlist doesn't have to be absolute it should be `path/to/wordlist`.

Author: addressed in 4fe26e86dfcc04de1115ef613c87693a5e0b7333

- Host-header fuzzing with host file on a traget website and match HTTP 200 responses:

> `ffuf -w {{hosts.txt}} -u {{https://example.org/}} -H "{{Host: FUZZ}}" -mc {{200}}`

Reviewer: ```suggestion
`ffuf -w {{path/to/hosts.txt}} -u {{https://example.org/}} -H "{{Host: FUZZ}}" -mc {{200}}`
```
Is the slash after `example.org` required in this case? If not I'd remove it.

Author: addressed in 4fe26e86dfcc04de1115ef613c87693a5e0b7333

# ffuf

> Subdomain and directory discovery tool.
> More information: <https://github.com/ffuf/ffuf>.

> - Discover directories using a wordlist on a target website colorized and verbose ouput:

Reviewer 2: I think we can highlight the argument meaning in the example description with the `[]`:
```suggestion
- Discover directories using a [w]ordlist on a target [u]rl with [c]olorized and [v]erbose ouput:
```
@marchersimon wdyt?

Author: I'm not against it, but I think that would be a paradigm shift for the TLDR project, as it is not currently implemented in other TLDR pages, i've come across.

Reviewer 2: @otterbaub we use there in some pages like `7z` and `tar` where we don't have long argument options and it's an easy way to let the user know what is each flag.

Author: Good enough for me, Convinced.

Addressed in a4924d75148c45de768d9fee1e6b4d83d36560ce

> - Discover directories using a [w]ordlist on a target website [c]olorized and [v]erbose ouput:

Reviewer 2: To match the `-u` flag:
```suggestion
- Discover directories using a [w]ordlist on a target [u]rl with [c]olorized and [v]erbose ouput:
```

> - Discover directories using a [w]ordlist on a target [u]rl with [c]olorized and [v]erbose ouput:

Reviewer 3: I didn't mean that there was any error with the letters for the flags being encased. I just personally dislike them, however if everyone else thinks they are fine then feel free to merge

Reviewer: I'm also not a fan, but I think that should be discussd in a dedicated issue.

Reviewer 2: @CleanMachine1 the PR got too long, the best was to merge it, and now we can refine the page.

`ffuf -w {{path/to/wordlist}} -u {{https://target/FUZZ}} -c -v`

> - Fuzz [h]ost-headers with host file on a traget website and [m]atch HTTP 200 [c]ode responses:

Reviewer 2: The flag is in upper case:
```suggestion
- Fuzz [H]ost-headers with host file on a traget website and [m]atch HTTP 200 [c]ode responses:
```

Author: good point, ty

> - Fuzz [H]ost-headers with a host file on a traget website and [m]atch HTTP 200 [c]ode responses:

Reviewer 3: Do we like all of the [H] in these?

I understand they signify the flags, but does it have long flags. 

PS: sorry if I have already asked this

Author: The ` [H] ` should be swapped as the flag is for "header" configuration and not "host" configuration.  
Good catch.

There are longer flags, but there are no instances of double tac(`--`) flags.
So there is `-H` and `-maxtime`.  There are no instances of `--Header` or `--MaximumTime`

addresed in 56d52cd8695b7ba70e0884664dff62a436c842c0

`ffuf -w {{hosts.txt}} -u {{https://example.org}} -H "{{Host: FUZZ}}" -mc {{200}}`

- Discover directory using a wordlist on a target website with a max individual job time of 60 seconds and recusrion discovery depth of 2 levels:

`ffuf -w {{path/to/wordlist}} -u {{https://target/FUZZ}} -maxtime-job {{60}} -recursion -recursion-depth {{2}}`

- Fuzz GET parameter on a target website and filter out message size response of 4242 bytes:

`ffuf -w {{path/to/param_names.txt}} -u {{https://target/script.php?FUZZ=test_value}} -fs {{4242}}`

- Fuzz POST method with POST data of password on a target website and filter out HTTP response code 401:

`ffuf -w {{path/to/postdata.txt}} -X {{POST}} -d "{{username=admin\&password=FUZZ}}" -u {{https://target/login.php}} -fc {{401}}`

- Discover subdomains using a subdoomain list on a target website:

> `ffuf -w {{subdomains.txt}} -u {{http://website.com}} -H {{“Host: FUZZ.website.com”}}`

Reviewer 2: The quotes used here are not the correct ones, and should be outside token syntax:
```suggestion
`ffuf -w {{subdomains.txt}} -u {{http://website.com}} -H "{{Host: FUZZ.website.com}}"`
```

Reviewer 3: How does one even get that character?

Author: addressed in 9127fcb4e00a7266c0fc853e19d80bfc0299ce95

> `ffuf -w {{subdomains.txt}} -u {{http://website.com}} -H "{{Host: FUZZ.website.com}}"`

Reviewer: ```suggestion
`ffuf -w {{subdomains.txt}} -u {{https://website.com}} -H "{{Host: FUZZ.website.com}}"`
```
Since we're using it everywhere else

Author: This brings up an interesting point.  HTTP vs. HTTPS.
The program can do both protocols non-secure and secure.  I went with examples using non-secure, as it is the simpler of the two.  That said, if we use HTTP/S in the examples, should there be a call out in the description of the program to inform that it is not protocol specific?
For example:
```
> Subdomain and directory discovery tool that uses HTTP or HTTP.
```
Thoughts?

Reviewer: I think if we specify `HTTPS` it's clear that the program can also handle `HTTP`, but maybe not otherwise.

Reviewer 3: Can we just drop it entirely?

```suggestion
`ffuf -w {{subdomains.txt}} -u {{website.com}} -H "{{Host: FUZZ.website.com}}"`
```

Author: The http/s is required for the program to work, other wise it will throw an error.

Reviewer 4: Let's change it to `https` as @marchersimon suggests.
